File: NC/src/CentralSecondDerivative.cpp
#include "../include/CentralSecondDerivative.hpp"

CentralSecondDerivative::CentralSecondDerivative(function<double(double)> f, double h)
    : Differentiation(f, h) {}

double CentralSecondDerivative::derivative(double x) const {
    return (func(x + h) - 2.0 * func(x) + func(x - h)) / (h * h);
}

File: Numerical_Computing_2/src/differentiation/central_second_derivative.py
from .numerical_differentiation import NumericalDifferentiation


class CentralSecondDerivative(NumericalDifferentiation):
    def derivative(self, x: float) -> float:
        return (
            self.func(x + self.h)
            - 2.0 * self.func(x)
            + self.func(x - self.h)
        ) / (self.h ** 2)

    @property
    def method_name(self) -> str:
        return "Central Difference (Second Derivative)"
